File: bindings/julia/test/schema_compatibility.jl
using Test

function python_kernel_schema_version()
    source_path = joinpath(@__DIR__, "..", "..", "..", "src", "innovate", "kernel.py")
    if !isfile(source_path) || get(ENV, "INNOVATE_JULIA_RUN_INSTALLED_PACKAGE_SMOKE", "false") == "true"
        return nothing
    end

    source = read(source_path, String)
    major = match(r"KERNEL_SCHEMA_MAJOR_VERSION\s*=\s*(\d+)", source)
    minor = match(r"KERNEL_SCHEMA_MINOR_VERSION\s*=\s*(\d+)", source)

    @test major !== nothing
    @test minor !== nothing

    return string(major.captures[1], ".", minor.captures[1])
end

@testset "Julia binding schema compatibility" begin
    python_schema_version = python_kernel_schema_version()

    @test Innovate.kernel_schema_version() == "1.0"
    @test isfile(Innovate.kernel_bridge_script())
    @test endswith(Innovate.kernel_bridge_script(), joinpath("inst", "python", "kernel_bridge.py"))

    if python_schema_version === nothing
        @test Innovate.kernel_request(operation = "discover_models")["schema_version"] == "1.0"
    else
        @test Innovate.kernel_schema_version() == python_schema_version
        @test Innovate.kernel_request(operation = "discover_models")["schema_version"] == python_schema_version
    end
end

File: bindings/rust/tests/schema_compatibility.rs
use std::fs;
use std::path::PathBuf;

use innovate_rust::{KernelBinding, KernelOperation, KernelRequest};

fn python_kernel_schema_version() -> String {
    let source_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("..")
        .join("..")
        .join("src")
        .join("innovate")
        .join("kernel.py");
    let source = fs::read_to_string(&source_path)
        .unwrap_or_else(|err| panic!("failed to read {}: {err}", source_path.display()));

    let major = source
        .lines()
        .find(|line| line.starts_with("KERNEL_SCHEMA_MAJOR_VERSION"))
        .and_then(|line| line.split('=').nth(1))
        .and_then(|value| value.trim().parse::<u32>().ok())
        .expect("expected KERNEL_SCHEMA_MAJOR_VERSION in Python kernel");

    let minor = source
        .lines()
        .find(|line| line.starts_with("KERNEL_SCHEMA_MINOR_VERSION"))
        .and_then(|line| line.split('=').nth(1))
        .and_then(|value| value.trim().parse::<u32>().ok())
        .expect("expected KERNEL_SCHEMA_MINOR_VERSION in Python kernel");

    format!("{major}.{minor}")
}

#[test]
fn schema_compatibility_guard_matches_the_python_kernel_contract() {
    let binding = KernelBinding::new();
    let python_schema_version = python_kernel_schema_version();

    assert_eq!(binding.schema_version(), python_schema_version);
    assert_eq!(
        KernelRequest::discover_models().schema_version,
        python_schema_version
    );

    let want_operations = [
        KernelOperation::DiscoverModels,
        KernelOperation::FitModel,
        KernelOperation::PredictModel,
        KernelOperation::SimulateModel,
        KernelOperation::SummarizeModel,
        KernelOperation::DiagnoseModel,
    ];
    assert_eq!(binding.available_operations(), want_operations);
    assert!(binding.bridge_script_exists());
    assert_eq!(
        want_operations
            .iter()
            .map(|operation| operation.as_str())
            .collect::<Vec<_>>(),
        vec![
            "discover_models",
            "fit_model",
            "predict_model",
            "simulate_model",
            "summarize_model",
            "diagnose_model",
        ]
    );
}
